common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -176,15 +176,31 @@ public Collation(
    */
 
   public static StringSearch getStringSearch(
-      final UTF8String left,
-      final UTF8String right,
+      final UTF8String targetUTF8String,
+      final UTF8String patternUTF8String,
       final int collationId) {
-    String pattern = right.toString();
-    CharacterIterator target = new StringCharacterIterator(left.toString());
+
+    if (collationId == UTF8_BINARY_COLLATION_ID) {
+      return getStringSearch(targetUTF8String, patternUTF8String);
+    } else if (collationId == UTF8_BINARY_LCASE_COLLATION_ID) {
+      return getStringSearch(targetUTF8String.toLowerCase(), patternUTF8String.toLowerCase());
+    }
+
+    String pattern = patternUTF8String.toString();
```

Comment:
General principle is that we should minimize heap allocations in string compare path.
In this snipped we are doing:
1) 2 allocations for UTF8_BINARY (should be 0)
2) For UTF8_BINARY_LCASE it get's kind of crazy. If this string has characters in non-ascii range we first push it to String, then back to UTF8String, then back to String :)

Reply:
It should be fine to do all the routing at compile time (e.g. in analysis/code gen phase). But in execution phase we should be aware of perf consequences.

Reply:
1. _2 allocations for UTF8_BINARY (should be 0)_ - this should be never be called for UTF8_BINARY
2. For this one, I understand the consequences, but it's very similar to what we had to do in the `UTF8String` to be able to successfully work with UTF8_BINARY_LCASE collation. This makes the code a lot cleaner than it was before when we had separate methods with _almost identical_ code for UTF8_BINARY_LCASE and other collations.

If this is not good/performant enough, we should think of a some other way to solve it because more and more PRs are coming with this change